Add character concept to ASCIICType functions

colelao opened the pull request:
#### 05263edf046c8c505a300e3656095b2452999f48
<pre>
Add character concept to ASCIICType functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=297085">https://bugs.webkit.org/show_bug.cgi?id=297085</a>
<a href="https://rdar.apple.com/157789937">rdar://157789937</a>

Reviewed by NOBODY (OOPS!).

This patch adds the concept of a character to ASCIICType.h. This will create better error messages for the developer when passing the wrong type. The functions will now accept only types that are convertible to a character.

* Source/WTF/wtf/ASCIICType.h:
(WTF::isASCII):
(WTF::isASCIILower):
(WTF::toASCIILowerUnchecked):
(WTF::isASCIIAlpha):
(WTF::isASCIIDigit):
(WTF::isASCIIAlphanumeric):
(WTF::isASCIIHexDigit):
(WTF::isASCIIBinaryDigit):
(WTF::isASCIIOctalDigit):
(WTF::isASCIIPrintable):
(WTF::isASCIIGraphic):
(WTF::isTabOrSpace):
(WTF::isASCIIWhitespace):
(WTF::isASCIIWhitespaceWithoutFF):
(WTF::isUnicodeCompatibleASCIIWhitespace):
(WTF::isASCIIUpper):
(WTF::isNotASCIIWhitespace):
(WTF::toASCIILower):
(WTF::toASCIIUpper):
(WTF::toASCIIHexValue):
(WTF::isASCIIAlphaCaselessEqual):
(WTF::isASCIIDigitOrPunctuation):
</pre>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f60d0f31d3422211cc08a27ad3ccac4b1a6908e3. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/49087

rreno commented on Source/WTF/wtf/ASCIICType.h:
```diff
@@ -39,57 +40,59 @@
-template<typename CharacterType> constexpr bool isUnicodeCompatibleASCIIWhitespace(CharacterType);
-template<typename CharacterType> constexpr bool isASCIIUpper(CharacterType);
+template<typename T> concept Character =
+    std::convertible_to<std::remove_cv_t<T>, char>;
```
style nit: this should probably be on one line

colelao committed:
Add character concept to ASCIICType functions
https://bugs.webkit.org/show_bug.cgi?id=297085
rdar://157789937

Reviewed by Ryan Reno.

This patch adds the concept of a character to ASCIICType.h. This will create better error messages for the developer when passing the wrong type. The functions will now accept only types that are convertible to a character.

* Source/WTF/wtf/ASCIICType.h:
(WTF::isASCII):
(WTF::isASCIILower):
(WTF::toASCIILowerUnchecked):
(WTF::isASCIIAlpha):
(WTF::isASCIIDigit):
(WTF::isASCIIAlphanumeric):
(WTF::isASCIIHexDigit):
(WTF::isASCIIBinaryDigit):
(WTF::isASCIIOctalDigit):
(WTF::isASCIIPrintable):
(WTF::isTabOrSpace):
(WTF::isASCIIWhitespace):
(WTF::isASCIIWhitespaceWithoutFF):
(WTF::isUnicodeCompatibleASCIIWhitespace):
(WTF::isASCIIUpper):
(WTF::isNotASCIIWhitespace):
(WTF::toASCIILower):
(WTF::toASCIIUpper):
(WTF::toASCIIHexValue):
(WTF::isASCIIAlphaCaselessEqual):
(WTF::isASCIIDigitOrPunctuation):

Canonical link: https://commits.webkit.org/298582@main
